[DO NOT MERGE] ADD improved spark submit cluster mode

ghost opened the pull request.

ajbosco committed:
[AIRFLOW-868] Add postgres_to_gcs operator and unittests

Adds a postgres_to_gcs operator to contrib so that a user can copy a
dump from postgres to google cloud storage. Tests write to local
NamedTemporayFiles so we correctly test serializing encoded ndjson in
both python3 and python2.7.

bolkedebruin committed:
[AIRFLOW-1802] Convert database fields to timezone aware

bolkedebruin committed:
[AIRFLOW-1804] Add time zone configuration options

Time zone defaults to UTC as is the default now in order
to maintain backwards compatibility.

bolkedebruin committed:
[AIRFLOW-1807] Force use of time zone aware db fields

This change will check if all date times being stored are
indeed timezone aware.

bolkedebruin committed:
[AIRFLOW-1806] Use naive datetime for cron scheduling

Converting to naive time is required in order to make sure
to run at exact times for crons.
E.g. if you specify to run at 8:00pm every day you do not
want suddenly to run at 7:00pm due to DST.

x committed:
[AIRFLOW-1613] make mysql_to_gcs_operator py3 compatible

Uses `__future__.unicode_literals` and replaces calling `json.dumps`
with `json.dump` followed by `tmp_file_handle.write` to write json lines
to the ndjson file. When using python3, `json.dump` will return a
unicode string instead of a byte string, therefore we encode the unicode
string to `utf-8` which is compatible with bigquery (see:
https://cloud.google.com/bigquery/docs/loading-data#loading_encoded_data).

bolkedebruin committed:
[AIRFLOW-1806] Use naive datetime when using cron

bolkedebruin committed:
[AIRFLOW-1827] Fix api endpoint date parsing

bolkedebruin committed:
[AIRFLOW-1826] Update views to use timezone aware objects

bolkedebruin committed:
[AIRFLOW-1803] Time zone documentation

bolkedebruin committed:
[AIRFLOW-1809] Update tests to use timezone aware objects

Commit:
ADD improved spark submit cluster mode

StephanErb committed:
[AIRFLOW-1559] Make database pooling optional

In situations where a database is heavily loaded with connections it
can be beneficial for operators to (temporarily) reduce the connection
footprint of Airflow on the database. This is particularly important
when Airflow or self-made extensions do not dispose the connection
pool when terminating.

Disabling the connection pool comes with a slowdown but that may be
acceptable in many deployment scenarios.

StephanErb committed:
[AIRFLOW-1559] Close file handles in subprocesses

All file descriptors except 0, 1 and 2 will be closed before the
child process is executed. This is the default on Python 3.2 and
above. This patch ensures consistent behaviour for older Python
versions.

Resources will be released once the main thread disposes
them, independent of the longevity of its subprocesses.

Background information:

* https://www.python.org/dev/peps/pep-0446/
* https://bugs.python.org/issue7213

StephanErb committed:
[AIRFLOW-1559] Dispose SQLAlchemy engines on exit

When a forked process or the entire interpreter terminates, we have
to close all pooled database connections. The database can run out
of connections otherwise. At a minimum, it will print errors in its
log file.

By using an atexit handler we ensure that connections are closed
for each interpreter and Gunicorn worker termination. Only usages
of multiprocessing.Process require special handling as those
terminate via os._exit() which does not run finalizers.

This commit is based on a contribution by @dhuang
https://github.com/apache/incubator-airflow/pull/2767

StephanErb committed:
[AIRFLOW-1665] Reconnect on database errors

This change enables the scheduler to recover from temporary database
errors and downtimes. The same holds true for the webserver if run
without its regular worker refresh.

The reconnect logic is based on a truncated exponential binary backoff
to ensure reconnect attempts don't overload the database.

Included changes:

* Switch to recommended pessimistic disconnect handling for engines
  http://docs.sqlalchemy.org/en/rel_1_1/core/pooling.html#disconnect-handling-pessimistic
* Remove legacy pool-based disconnect handling.
* Ensure event handlers are registered for each newly created engine.
  Engines are re-initialized in child processes so this is crucial for
  correctness.

This commit is based on a contribution by @vklogin
https://github.com/apache/incubator-airflow/pull/2744

bolkedebruin committed:
[AIRFLOW-1808] Convert all utcnow() to time zone aware

datetime.utcnow() does not set time zone information.

sv3ndk committed:
various improvements to log processing

Commit:
[AIRFLOW-1843] Add Google Cloud Storage Sensor with prefix

Sensor for checking if there any files in bucket
at certain prefix

Closes #2809 from litdeviant/gcs_prefix_sensor

Commit:
[AIRFLOW-1848][Airflow-1848] Fix DataFlowPythonOperator py_file extension doc comment

Closes #2816 from cjqian/1848

ghost commented:
https://issues.apache.org/jira/browse/AIRFLOW-1854

Commit:
ADD kill command IMPROVE comments CLEANUP code

Commit:
ADD unit tests spark submit

Commit:
REMOVE parameter

Commit:
ADD if test

sv3ndk commented on airflow/contrib/hooks/spark_submit_hook.py:
```diff
@@ -116,6 +117,10 @@ def __init__(self,
         self._connection = self._resolve_connection()
         self._is_yarn = 'yarn' in self._connection['master']
 
+        self._track_driver_state = 'spark://' in self._connection['master'] and self._connection['deploy_mode'] == 'cluster'
```
sorry to be picky, how about renaming the variable `_track_driver_state` into `_should_track_driver_status`? 
Since there are no types in python, the starting "should" makes the code more obvious by letting everyone know this is a boolean, and `status` is better aligned with spark's vocabulary => this also makes the code easier to read for newcomers.

sv3ndk replied:
Also, could you please extract this into a method and add unit tests to it to check that we correctly return true and false for plenty of types of `connection` object? 
I think this is important to convince the airflow community that we are not breaking anything here, since the new mechanism we're introduction must not trigger for all the (client mode) cases for which they are using it => such unit -test is a way of demonstrating that

sv3ndk commented on airflow/contrib/hooks/spark_submit_hook.py:
```diff
@@ -220,6 +225,31 @@ def _build_command(self, application):
 
         return connection_cmd
 
+    def _build_track_driver_state_command(self):
```
(idem: let's rename "state" into "status" here)

sv3ndk commented on airflow/contrib/hooks/spark_submit_hook.py:
```diff
@@ -262,16 +315,115 @@ def _process_log(self, itr):
+            connection_cmd = [self._connection['spark_binary']]
+
+        # The url ot the spark master
+        connection_cmd += ["--master", self._connection['master']]
```
this if/else logic to build the beginning of the spark submit `connection_cmd` is duplicated in the code, I suggest we refactor it to have it in only place.

sv3ndk reviewed (changes requested):
great job :) 
Some minor comments on wording tests

Commit:
[AIRFLOW-1850] Copy cmd before masking

The cmd is first copied before the password is
masked. This ensures
that the orignal cmd isn't changed. Replacing the
password with a
masked value replaces the password in the original
command since it
is passed by reference.

Closes #2817 from Fokko/AIRFLOW-1850-copy-cmd-
before-replacing-password

Commit:
[AIRFLOW-1785] Enable Python 3 tests

Enable tests under Python 3 to make sure that
tests run under Python
3.

Closes #2755 from Fokko/AIRFLOW-1785-Enable-
Python3-tests

bolkedebruin committed:
[AIRFLOW-1870] Enable flake8 tests

Flake8 tests now run for diffs

Closes #2829 from bolkedebruin/use_flake8

kaxil committed:
[AIRFLOW-1855][AIRFLOW-1866] Add GCS Copy Operator to copy multiple files

Closes #2819 from kaxil/master

bolkedebruin committed:
[AIRFLOW-1872] Set context for all handlers including parents

Previously setting the context was not propagated
to the parent
loggers. Unfortnately, in case of a non explicitly
defined logger
the returned logger is shallow, ie. it does not
have handlers
defined. So to set the context it is required to
walk the tree.

Closes #2831 from bolkedebruin/fix_logging

Commit:
CLEANUP code ADD tests for track_driver_status

kaxil committed:
[AIRFLOW-1883] Get File Size for objects in Google Cloud Storage

Closes #2840 from kaxil/Get_File_Size

kaxil committed:
[AIRFLOW-XXX] Added DataReply to the list of Airflow Users

Closes #2841 from kaxil/patch-1

bolkedebruin committed:
[AIRFLOW-1881] Make operator log in task log

Previously operators logged under
airflow.operators or
airflow.contrib.operators. This unifies them under
airflow.task.operators allowing the task log to
pick
them up and not have 'double' logging.

Closes #2838 from bolkedebruin/AIRFLOW-1881

bolkedebruin committed:
[AIRFLOW-966] Make celery broker_transport_options configurable

Required for changing visibility timeout and other
options required
for Redis/SQS.

Closes #2842 from bolkedebruin/AIRFLOW-966

bolkedebruin committed:
[AIRFLOW-342] Do not use amqp, rpc as result backend

amqp and rpc (and redis most likely) cannot store
results for tasks
long enough.

Closes #2830 from bolkedebruin/AIRFLOW-342

Commit:
FIX redundant boolean checks FIX sp.wait

Commit:
REMOVE on_kill logic

alexisrosuel committed:
add hostnfly as users of airflow

Closes #2845 from alexisrosuel/master

pdambrauskas committed:
[AIRFLOW-1554] Fix wrong DagFileProcessor termination method call

Closes #2821 from
pdambrauskas/fix/wrong_termination_call

Commit:
[AIRFLOW-1876] Write subtask id to task log header

Closes #2835 from wrp/subtask-id

Commit:
[AIRFLOW-1869] Write more error messages into gcs and file logs

Closes #2826 from wrp/gcs-log

bolkedebruin committed:
[AIRFLOW-1879] Handle ti log entirely within ti

Previously logging was setup outside a
TaskInstance,
this puts everything inside. Also propery closes
the logging.

Closes #2837 from bolkedebruin/AIRFLOW-1879

Commit:
Update README.md

Added Playbuzz to list of companies using Airflow.

Closes #2828 from clintonboys/patch-5

sv3ndk commented on airflow/contrib/hooks/spark_submit_hook.py:
```diff
@@ -110,12 +111,19 @@ def __init__(self,
+        self._driver_id = None
+        self._driver_status = None
+
+    def _resolve_should_track_driver_status(self):
```
how about adding some doc string here? something like 

"determines whether or not this hook should poll the spark driver status through subsequent spark-submit requests after the initial spark-submit request"

sv3ndk commented on airflow/contrib/hooks/spark_submit_hook.py:
```diff
@@ -220,6 +233,25 @@ def _build_command(self, application):
+        # The driver id so we can poll for its status
+        if self._driver_id:
+            connection_cmd += ["--status", self._driver_id]
+
```
I understand there is already a first check in `_process_log` though I would also raise an exception here in the `else` case, something like
```
else
    raise AirflowException("Invalid status: attempted to poll driver status but no driver id is known. Giving up.")
```

Otherwise, in case of bug where the driver_id is not known, this function would silently create a weird to interpret error due to incorrect spark-submit process

sv3ndk commented on airflow/contrib/hooks/spark_submit_hook.py:
```diff
@@ -246,9 +278,32 @@ def submit(self, application="", **kwargs):
+        Processes the log files and extracts useful information out of it.
+
+        Remark: If the driver needs to be tracked for its status, the log-level of the spark deploy needs to be
+        at least INFO (log4j.logger.org.apache.spark.deploy.rest=INFO)
```
I updated the log4j again the other day, I now think it's  even better to log like this: 
```
log4j.logger.org.apache.spark.deploy=INFO
```

sv3ndk commented on airflow/contrib/hooks/spark_submit_hook.py:
```diff
@@ -262,16 +317,82 @@ def _process_log(self, itr):
+                    self._driver_id = match_driver_id.groups()[0]
+                    self.log.info("identified spark driver id: {}".format(self._driver_id))
+
+            self.log.debug("spark submit log: {}".format(line))
```
Could you put that status as `self.log.info` ? I find it important to always have the spark submit statement in the logs. I know it was originally in debug in the original version, though I suggest to update that. 
(same thing for the same log in `_process_spark_status_log`)

milanvdmria replied:
I dont agree as it creates a lot of noise IMO.
The printed lines are also not relevant for day to day operation.
If something goes wrong, then you can start Airflow in debug mode and those lines will show up.

sv3ndk commented on airflow/contrib/hooks/spark_submit_hook.py:
```diff
@@ -262,16 +317,82 @@ def _process_log(self, itr):
+                    "Failed to poll for the driver status: returncode = {}".format(returncode)
+                    )
+
+            # Sleep for 1 second as we do not want to spam the cluster
```
If we move this sleep at the beginning of the `while` loop, we'll avoid waiting one last time after the batch is finished => we'll be slightly faster end-to-end :)

sv3ndk commented on airflow/contrib/operators/spark_submit_operator.py:
```diff
@@ -145,7 +145,7 @@ def execute(self, context):
             num_executors=self._num_executors,
             application_args=self._application_args,
-            verbose=self._verbose
+            verbose=self._verbose,
```
Could you remove the coma here? Sorry, I'm picky, it just make the perceived PR size smaller by removing one "modified" file

sv3ndk reviewed (changes requested):
Awesome Milan, I really like it :) 
I put tons of comment as an attempt the "beautify" the PR as much as possible in order to minimise the friction to get it merged.

Commit:
IMPROVEMENTS

Commit:
[AIRFLOW-1854] Improve Spark Submit operator for standalone cluster mode
